sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -375,4 +422,361 @@ public void checkDone() throws IllegalStateException {
+        restrictionTracker(
+            @Restriction KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT> restriction,
+            PipelineOptions pipelineOptions) {
+      return new UnboundedSourceAsSDFRestrictionTracker(restriction, pipelineOptions);
+    }
+
+    @ProcessElement
+    public ProcessContinuation processElement(
+        RestrictionTracker<
+                KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT>, UnboundedSourceValue[]>
+            tracker,
+        OutputReceiver<ValueWithRecordId<OutputT>> receiver,
+        BundleFinalizer bundleFinalizer)
+        throws IOException {
+      UnboundedSourceValue<OutputT>[] out = new UnboundedSourceValue[1];
+      while (tracker.tryClaim(out)) {
+        receiver.outputWithTimestamp(
+            new ValueWithRecordId<>(out[0].getValue(), out[0].getId()), out[0].getTimestamp());
+      }
+
+      // Add the checkpoint mark to be finalized if the checkpoint mark isn't trivial.
+      KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT> currentRestriction =
+          tracker.currentRestriction();
+      if (currentRestriction.getValue() != null
+          && !(tracker.currentRestriction().getValue() instanceof NoopCheckpointMark)) {
+        bundleFinalizer.afterBundleCommit(
+            Instant.now().plus(Duration.standardMinutes(DEFAULT_BUNDLE_FINALIZATION_LIMIT_MINS)),
+            currentRestriction.getValue()::finalizeCheckpoint);
+      }
+
+      // If we have been split/checkpoint by a runner, the tracker will have been updated to the
+      // empty source and we will return stop. Otherwise the unbounded source has only temporarily
+      // run out of work.
+      if (tracker.currentRestriction().getKey() instanceof EmptyUnboundedSource) {
+        return ProcessContinuation.stop();
+      }
+      return ProcessContinuation.resume();
+    }
+
+    @GetRestrictionCoder
+    public Coder<KV<UnboundedSource<OutputT, CheckpointT>, CheckpointT>> restrictionCoder() {
+      return KvCoder.of(
+          SerializableCoder.of(new TypeDescriptor<UnboundedSource<OutputT, CheckpointT>>() {}),
+          NullableCoder.of(restrictionCoder));
+    }
+
+    /**
+     * A named tuple representing all the values we need to pass between the {@link UnboundedReader}
+     * and the {@link org.apache.beam.sdk.transforms.DoFn.ProcessElement @ProcessElement} method of
+     * the splittable DoFn.
+     */
+    @AutoValue
+    abstract static class UnboundedSourceValue<T> {
+      public static <T> UnboundedSourceValue<T> create(
+          byte[] id, T value, Instant timestamp, Instant watermark) {
+        return new AutoValue_Read_UnboundedSourceAsSDFWrapperFn_UnboundedSourceValue<T>(
+            id, value, timestamp, watermark);
+      }
+
+      @SuppressWarnings("mutable")
+      public abstract byte[] getId();
+
+      public abstract T getValue();
+
+      public abstract Instant getTimestamp();
+
+      public abstract Instant getWatermark();
```

Comment:
Also curious why each output has its own `watermark`. Will the `watermark` be different from `timestamp` for one given record?

Reply:
The UnboundedSource can report a watermark that is reported separately from the element timestamp.